File: population.rb
def read_rows(input)
  return to_enum(:read_rows, input) unless block_given?
  input.each do |line|
    yield line.split("\t")
  end
end

def calc(input)
  population_south = 0
  population_total = 0
  row_count = 0
  rows = read_rows(input)
  # rows.each do |row|
  tally = proc do |row|
  # read_rows(input) do |row|
    latitude = row[4].to_f
    population = row[14].to_i
    population_total += population
    if latitude < 0
      population_south += population
    end
    row_count += 1
  end
  rows.each do |row|
    tally.call(row)
    break if population_total > 1e9
  end
  puts "Reached a billion!"
  rows.each do |row|
    tally.call(row)
  end
  puts "South population: #{population_south}"
  puts "Total population: #{population_total}"
  puts "# of rows: #{row_count}"
end

def read_rows_fiber(input)
  Fiber.new do
    input.each do |line|
      Fiber.yield line.split("\t")
    end
  end
end

def step(input)
  # rows = read_rows_fiber(input)
  # first = rows.resume
  rows = read_rows(input)
  first = rows.next
  puts "lat #{first[4]} with pop #{first[14]}"
end

File.open("data/cities500.txt") do |input|
  calc(input)
  # step(input)
end

# [1, 2, 3].each do |i|
#   puts i
#   break
# end

# numbers = [1, 2, 3].each
# puts numbers.next

File: population.py
# Amortize the cost of importing pandas.
from pandas import read_csv


def read_rows(in_stream):
    for line in in_stream:
        yield line.split('\t')


def count_gen():
    with open('data/cities500.txt') as in_stream:
        population_south = 0
        population_total = 0
        row_count = 0
        for row in read_rows(in_stream):
            latitude = float(row[4])
            population = int(row[14])
            population_total += population
            if latitude < 0:
                population_south += population
            row_count += 1
        return dict(
            population_south=population_south,
            population_total=population_total,
            row_count=row_count)


def count_pandas():
    # from pandas import read_csv
    fields = dict(lat=4, pop=14)
    cities = read_csv(
        'data/cities500.txt', header=None, sep='\t',
        names=fields.keys(),
        usecols=fields.values())
    return dict(
        population_south=cities[cities['lat'] < 0]['pop'].sum(),
        population_total=cities['pop'].sum(),
        row_count=cities.shape[0])


def main():
    from sys import argv
    mode = dict(gen=count_gen, pandas=count_pandas)
    result = mode.get(argv[1])()
    print(f'South population: {result["population_south"]}')
    print(f'Total population: {result["population_total"]}')
    print(f'# of rows: {result["row_count"]}')


if __name__ == "__main__":
    main()
